// ngx-custom-validators/esm5/src/app/min-date/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, Input, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { minDate } from './validator';
/** @type {?} */
var MIN_DATE_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef(function () { return MinDateValidator; }),
    multi: true
};
var MinDateValidator = /** @class */ (function () {
    function MinDateValidator() {
    }
    /**
     * @return {?}
     */
    MinDateValidator.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.validator = minDate(this.minDate);
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    MinDateValidator.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        for (var key in changes) {
            if (key === 'minDate') {
                this.validator = minDate(changes[key].currentValue);
                if (this.onChange) {
                    this.onChange();
                }
            }
        }
    };
    /**
     * @param {?} c
     * @return {?}
     */
    MinDateValidator.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return this.validator(c);
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    MinDateValidator.prototype.registerOnValidatorChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    MinDateValidator.decorators = [
        { type: Directive, args: [{
                    selector: '[minDate][formControlName],[minDate][formControl],[minDate][ngModel]',
                    providers: [MIN_DATE_VALIDATOR]
                },] }
    ];
    MinDateValidator.propDecorators = {
        minDate: [{ type: Input }]
    };
    return MinDateValidator;
}());
export { MinDateValidator };
if (false) {
    /** @type {?} */
    MinDateValidator.prototype.minDate;
    /** @type {?} */
    MinDateValidator.prototype.validator;
    /** @type {?} */
    MinDateValidator.prototype.onChange;
}
