// ngx-custom-validators/esm5/src/app/greater-than-equal/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, Input, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { gte } from './validator';
/** @type {?} */
var GREATER_THAN_EQUAL_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef(function () { return GreaterThanEqualValidator; }),
    multi: true
};
var GreaterThanEqualValidator = /** @class */ (function () {
    function GreaterThanEqualValidator() {
    }
    /**
     * @return {?}
     */
    GreaterThanEqualValidator.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.validator = gte(this.gte);
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    GreaterThanEqualValidator.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        for (var key in changes) {
            if (key === 'gte') {
                this.validator = gte(changes[key].currentValue);
                if (this.onChange) {
                    this.onChange();
                }
            }
        }
    };
    /**
     * @param {?} c
     * @return {?}
     */
    GreaterThanEqualValidator.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return this.validator(c);
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    GreaterThanEqualValidator.prototype.registerOnValidatorChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    GreaterThanEqualValidator.decorators = [
        { type: Directive, args: [{
                    selector: '[gte][formControlName],[gte][formControl],[gte][ngModel]',
                    providers: [GREATER_THAN_EQUAL_VALIDATOR]
                },] }
    ];
    GreaterThanEqualValidator.propDecorators = {
        gte: [{ type: Input }]
    };
    return GreaterThanEqualValidator;
}());
export { GreaterThanEqualValidator };
if (false) {
    /** @type {?} */
    GreaterThanEqualValidator.prototype.gte;
    /** @type {?} */
    GreaterThanEqualValidator.prototype.validator;
    /** @type {?} */
    GreaterThanEqualValidator.prototype.onChange;
}
